common/client/versionChecker.go

```diff
@@ -0,0 +1,165 @@
+	}
+)
+
+// NewVersionChecker constructs a new VersionChecker
+func NewVersionChecker() VersionChecker {
+	supportedFeatures := map[string]map[string]version.Constraints{
+		GoSDK: {
+			stickyQuery:     mustNewConstraint(fmt.Sprintf(">=%v", GoWorkerStickyQueryVersion)),
+			consistentQuery: mustNewConstraint(fmt.Sprintf(">=%v", GoWorkerConsistentQueryVersion)),
+		},
+		JavaSDK: {
+			stickyQuery: mustNewConstraint(fmt.Sprintf(">=%v", JavaWorkerStickyQueryVersion)),
+		},
+	}
+	supportedClients := map[string]version.Constraints{
+		GoSDK:   mustNewConstraint(fmt.Sprintf("<=%v", SupportedGoSDKVersion)),
+		JavaSDK: mustNewConstraint(fmt.Sprintf("<=%v", SupportedJavaSDKVersion)),
+		CLI:     mustNewConstraint(fmt.Sprintf("<=%v", SupportedCLIVersion)),
+	}
+	return &versionChecker{
+		supportedFeatures: supportedFeatures,
+		supportedClients:  supportedClients,
+	}
+}
+
+// ClientSupported returns an error if client is unsupported, nil otherwise.
+// In case client version lookup fails assume the client is supported.
+func (vc *versionChecker) ClientSupported(ctx context.Context, enableClientVersionCheck bool) error {
+	if !enableClientVersionCheck {
+		return nil
+	}
+
+	call := yarpc.CallFromContext(ctx)
+	clientFeatureVersion := call.Header(common.FeatureVersionHeaderName)
+	clientImpl := call.Header(common.ClientImplHeaderName)
+
+	if clientFeatureVersion == "" {
+		return nil
+	}
+	supportedVersions, ok := vc.supportedClients[clientImpl]
+	if !ok {
+		return nil
+	}
+	version, err := version.NewVersion(clientFeatureVersion)
+	if err != nil {
+		return &shared.ClientVersionNotSupportedError{FeatureVersion: clientFeatureVersion, ClientImpl: clientImpl, SupportedVersions: supportedVersions.String()}
+	}
+	if !supportedVersions.Check(version) {
+		return &shared.ClientVersionNotSupportedError{FeatureVersion: clientFeatureVersion, ClientImpl: clientImpl, SupportedVersions: supportedVersions.String()}
+	}
+	return nil
+}
+
+// SupportsStickyQuery returns error if sticky query is not supported otherwise nil.
+// In case client version lookup fails assume the client does not support feature.
+func (vc *versionChecker) SupportsStickyQuery(clientImpl string, clientFeatureVersion string) error {
+	return vc.featureSupported(clientImpl, clientFeatureVersion, stickyQuery)
+}
+
+// SupportsConsistentQuery returns error if consistent query is not supported otherwise nil.
+// In case client version lookup fails assume the client does not support feature.
+func (vc *versionChecker) SupportsConsistentQuery(clientImpl string, clientFeatureVersion string) error {
+	return vc.featureSupported(clientImpl, clientFeatureVersion, consistentQuery)
+}
+
+func (vc *versionChecker) featureSupported(clientImpl string, clientFeatureVersion string, feature string) error {
+	if clientImpl == "" || clientFeatureVersion == "" {
+		return &shared.ClientVersionNotSupportedError{ClientImpl: clientImpl, FeatureVersion: clientFeatureVersion}
```

Comment:
I don't think so for two reasons:

- The customer will get this error in response to a feature usage. For example they will call query and then get this error back so its clear that the error has to do with query.
- Second this error does not always relate to features, it can also just mean the client is too new for this server in which case there is no specific feature.